src/dust/dust.cpp

```diff
@@ -320,6 +339,379 @@ TaskStatus FluxSource(MeshData<Real> *md, const Real dt) {
+template <Coordinates GEOM>
+TaskCollection OperatorSplitDustSelect(Mesh *pm, parthenon::SimTime &tm) {
+  TaskCollection tc;
+  auto &coag_pkg = pm->packages.Get("coagulation");
+  auto *dtCoag = coag_pkg->MutableParam<Real>("dtCoag");
+  int nstep1Coag = coag_pkg->template Param<int>("nstep1Coag");
+  if (tm.ncycle == 0) {
+    *dtCoag = 0.0;
+  }
+
+  *dtCoag += tm.dt;
+  if ((tm.ncycle + 1) % nstep1Coag != 0) {
+    return tc;
+  }
+
+  const Real time_local = tm.time + tm.dt - (*dtCoag);
+  const Real dt_local = (*dtCoag);
+  if (parthenon::Globals::my_rank == 0) {
+    std::cout << "coagulation at: time,dt,cycle=" << time_local << " " << dt_local << " "
+              << tm.ncycle << std::endl;
+  }
+
+  TaskID none(0);
+
+  // Assemble tasks
+  auto &coag = coag_pkg->template Param<Dust::Coagulation::CoagParams>("coag_pars");
+  const int nspecies = coag.nm;
+  std::vector<std::string> dust_var_names;
+  for (int n = 0; n < nspecies; n++) {
+    dust_var_names.push_back(dust::prim::density::name() + '_' + std::to_string(n));
+    dust_var_names.push_back(dust::prim::velocity::name() + '_' + std::to_string(n));
+  }
+  auto &dust_subset =
+      pm->mesh_data.AddShallow("dust_subset", pm->mesh_data.Get(), dust_var_names);
+
+  using namespace ::parthenon::Update;
+  const int num_partitions = pm->DefaultNumPartitions();
+  TaskRegion &tr = tc.AddRegion(num_partitions);
+  for (int i = 0; i < num_partitions; i++) {
+    auto &tl = tr[i];
+    auto &base = pm->mesh_data.GetOrAdd("base", i);
+    auto coag_step =
+        tl.AddTask(none, CoagulationOneStep<GEOM>, base.get(), time_local, dt_local);
+
+    // Set (remaining) fields to be communicated
+    auto pre_comm = tl.AddTask(coag_step, PreCommFillDerived<MeshData<Real>>, base.get());
+
+    // Set boundary conditions (both physical and logical)
+    // auto bcs = parthenon::AddBoundaryExchangeTasks(pre_comm, tl, base, pm->multilevel);
+    auto &md_coag = pm->mesh_data.GetOrAdd("dust_subset", i);
+    auto bcs = parthenon::AddBoundaryExchangeTasks(pre_comm, tl, md_coag, pm->multilevel);
+
+    // Sync fields
+    auto p2c = tl.AddTask(bcs, FillDerived<MeshData<Real>>, base.get());
+  }
+  *dtCoag = 0.0;
+
+  return tc;
+}
+
+//----------------------------------------------------------------------------------------
+//! \fn  TaskStatus Dust::OperatorSplit
+//  \brief Wrapper function for Dust::OperatorSplitDust
+TaskListStatus OperatorSplitDust(Mesh *pm, parthenon::SimTime &tm) {
+  auto &dust_pkg = pm->packages.Get("dust");
+  typedef Coordinates C;
+  const C coords = dust_pkg->template Param<Coordinates>("coords");
+
+  if (coords == C::cartesian) {
+    return OperatorSplitDustSelect<C::cartesian>(pm, tm).Execute();
+  } else if (coords == C::spherical1D) {
+    return OperatorSplitDustSelect<C::spherical1D>(pm, tm).Execute();
+  } else if (coords == C::spherical2D) {
+    return OperatorSplitDustSelect<C::spherical2D>(pm, tm).Execute();
```

Comment:
`OperatorSplitDust` is the function that I'm suggesting be put into the driver class.  You then have access to `SimTime` and other things. You can also template it on the coords, so you don't need `OperatorSplitDustSelect` at all.